src/types.jl

```diff
@@ -0,0 +1,200 @@
+## We have different types:
+##
+## Basic: holds a ptr to a symengine object. Faster, so is default type
+##
+## BasicType{Val{:XXX}}: types that can be use to control dispatch
+##
+## SymbolicType: is a type union of the two
+##
+## Basic(x::BasicType) gives a basic object; BasicType(x::Basic) gives a BasicType object. (This name needs change)
+##
+## To control dispatch, one might have `N(b::Basic) = N(BasicType(b))` and then define `N` for types of interest
+
+## Hold a reference to a SymEngine object
+type Basic  <: Number
+    ptr::Ptr{Void}
+    function Basic()
+        z = new(C_NULL)
+        ccall((:basic_new_stack, :libsymengine), Void, (Ptr{Basic}, ), &z)
+        finalizer(z, basic_free)
+        return z
+    end
+end
+export Basic
+
+
+basic_free(b::Basic) = ccall((:basic_free_stack, :libsymengine), Void, (Ptr{Basic}, ), &b)
+
+
+
+function Basic(x::Clong)
+    a = Basic()
+    ccall((:integer_set_si, :libsymengine), Void, (Ptr{Basic}, Clong), &a, x)
+    return a
+end
+
+function Basic(x::Culong)
+    a = Basic()
+    ccall((:integer_set_ui, :libsymengine), Void, (Ptr{Basic}, Culong), &a, x)
+    return a
+end
+
+function Basic(x::BigInt)
+    a = Basic()
+    ccall((:integer_set_mpz, :libsymengine), Void, (Ptr{Basic}, Ptr{BigInt}), &a, &x)
+    return a
+end
+
+
+if Clong == Int32
+    convert(::Type{Basic}, x::Union{Int8, Int16, Int32}) = Basic(convert(Clong, x))
+    convert(::Type{Basic}, x::Union{UInt8, UInt16, UInt32}) = Basic(convert(Culong, x))
+else
+    convert(::Type{Basic}, x::Union{Int8, Int16, Int32, Int64}) = Basic(convert(Clong, x))
+    convert(::Type{Basic}, x::Union{UInt8, UInt16, UInt32, UInt64}) = Basic(convert(Culong, x))
+end
+convert(::Type{Basic}, x::Integer) = Basic(BigInt(x))
+convert(::Type{Basic}, x::Rational) = Basic(num(x)) / Basic(den(x))
+
+Base.promote_rule{S<:Number}(::Type{Basic}, ::Type{S} ) = Basic
+
+## Class ID
+get_type(s::Basic) = convert(Int, ccall((:basic_get_type, :libsymengine), UInt, (Ptr{Basic},), &s))
+get_class_from_id(id::Int) = ccall((:basic_get_class_from_id, :libsymengine), Ptr{UInt8}, (Int,), id) |> bytestring
```

Comment:
Thanks again!

On Sat, Apr 2, 2016 at 11:45 PM, Isuru Fernando <notifications@github.com>
wrote:

> In src/types.jl
> <https://github.com/symengine/SymEngine.jl/pull/20#discussion_r58304047>:
>
> > +
> > +if Clong == Int32
> > +    convert(::Type{Basic}, x::Union{Int8, Int16, Int32}) = Basic(convert(Clong, x))
> > +    convert(::Type{Basic}, x::Union{UInt8, UInt16, UInt32}) = Basic(convert(Culong, x))
> > +else
> > +    convert(::Type{Basic}, x::Union{Int8, Int16, Int32, Int64}) = Basic(convert(Clong, x))
> > +    convert(::Type{Basic}, x::Union{UInt8, UInt16, UInt32, UInt64}) = Basic(convert(Culong, x))
> > +end
> > +convert(::Type{Basic}, x::Integer) = Basic(BigInt(x))
> > +convert(::Type{Basic}, x::Rational) = Basic(num(x)) / Basic(den(x))
> > +
> > +Base.promote_rule{S<:Number}(::Type{Basic}, ::Type{S} ) = Basic
> > +
> > +## Class ID
> > +get_type(s::Basic) = convert(Int, ccall((:basic_get_type, :libsymengine), UInt, (Ptr{Basic},), &s))
> > +get_class_from_id(id::Int) = ccall((:basic_get_class_from_id, :libsymengine), Ptr{UInt8}, (Int,), id) |> bytestring
>
> a = ccall((:basic_get_class_from_id, :libsymengine), Ptr{UInt8}, (Int,), id)
> string = bytestring(a)
> ccall((:basic_str_free, :libsymengine), Void, (Ptr{UInt8}, ), a)
> string
>
> should fix the leak
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/symengine/SymEngine.jl/pull/20/files/4b995dc4bba02cf4263c3a9daf197928ce187b16#r58304047>
>



-- 
John Verzani
Chair, Department of Mathematics
College of Staten Island, CUNY
verzani@math.csi.cuny.edu